### ISSUE TITLE
Tesseract OCR: less memory consumption by avoiding new instances of TessBaseAPI?

### ISSUE THREAD

**Issue by DeborahAnn032** [adaptech-cz/Tesseract4Android#56]
Not really an issue but I thought about asking here since I don't know exactly how this API wrapper works.

Basically I make a new instance of MyOCR whenever I need to perform OCR.
This is currently what my constructor looks like:

    public MyOCR(Bitmap bitmap)
    	{
    		this.tessBaseAPI = new TessBaseAPI();
    		this.tessBaseAPI.setPageSegMode(TessBaseAPI.PageSegMode.PSM_AUTO_ONLY);
    		try {
    			this.tessBaseAPI.setDebug(true);
    			this.tessBaseAPI.init("storage/emulated/0", "eng");
    			this.tessBaseAPI.setImage(bitmap);
    			this.text = tessBaseAPI.getUTF8Text();
    			this.tessBaseAPI.end();
    		} catch (Exception e) {
    			e.printStackTrace();
    			System.err.println(e.getMessage());
    		}
    	}

I was wondering if performance wise, in the long run, the following code be preferable.
Basically I make only one instance of MyOCR and set the new image every time I need to perform OCR.

    public MyOCR()
    {
    	this.tessBaseAPI = new TessBaseAPI();
    	this.tessBaseAPI.setPageSegMode(TessBaseAPI.PageSegMode.PSM_AUTO_ONLY);
    	try {
    		this.tessBaseAPI.setDebug(true);
    		this.tessBaseAPI.init("storage/emulated/0", "eng");
    	} catch (Exception e) {
    		e.printStackTrace();
    		System.err.println(e.getMessage());
    	}
    }
    
    public void ocrTask(Bitmap bitmap)
    {
    	this.tessBaseAPI.setImage(bitmap);
    	this.text = tessBaseAPI.getUTF8Text();
    }

**Comment by Robyer** [comment]
Yes, your second example will be definitely faster and is preferred way in most situations.

There are multiple initialization "steps" which could be reused.

1) The native instance itself, which is created when you create the new TessBaseApi() instance. It is completely released only when you call recycle() on it.

2) Initialized engine with given traineddata by calling init(). You can call this again when you want to change the traineddata - without releasing the whole native instance (as in 1).

3) You can process different images by calling setImage() and getUTF8text() multiple times without reinitializing the engine (as in 2).

Some tesseract parameters can be also changed during the lifetime of the instance, but you would need to check the native source code to see how it affects it.

Additionally, I expect you came from using tess-two library, since you use end() method to release instance, but it wasn't working correctly, so I changed it to method release() to fully release the instance.

You can look at the sample project to understand how you can reuse the instance, together with progress listener and ability to stop the ongoing processing.

https://github.com/adaptech-cz/Tesseract4Android/blob/master/sample/src/main/java/cz/adaptech/tesseract4android/sample/ui/main/MainViewModel.java

**Comment by DeborahAnn032** [comment]
Thanks so much for the reply!

Yes, definitely more faster but I am more concerned with memory consumption because my app gets killed by the OS after a while, that's why I am not sure if it's better the first solution (in which it's realesed right  after the job it's done since it creates each time a new TessBaseAPI) or the second one (in which the release would be called only when the app is closed).

Yes, I was using tess-two, I will now make the proper adjustments and use recycle() instead of end().


By the way, why you call also stop() in the sample you provided?

**Comment by DeborahAnn032** [comment]
Thanks so much for all the explanations! Of course calling first clear() and right after recycle() would be redundant, right? If you guarantee me that recycle() frees up all memory I will call just that when I am done.

**Comment by Robyer** [comment]
Yes, recycle() releases everything (internally it also calls clear()). After that no method can be called on this instance anymore.
